ethcore/src/executive.rs

```diff
@@ -165,6 +189,559 @@ impl TransactOptions<trace::NoopTracer, trace::NoopVMTracer> {
+						Ok(()) => (),
+						Err(err) => return Ok(Err(err)),
+					}
+				}
+
+				let origin_info = OriginInfo::from(&params);
+				let exec = self.factory.create(params, self.schedule, self.depth);
+
+				let out = {
+					let mut ext = Self::as_externalities(state, self.info, self.machine, self.schedule, self.depth, self.stack_depth, self.static_flag, &origin_info, &mut unconfirmed_substate, OutputPolicy::InitContract, tracer, vm_tracer);
+					match exec.exec(&mut ext) {
+						Ok(val) => Ok(val.finalize(ext)),
+						Err(err) => Err(err),
+					}
+				};
+
+				let res = match out {
+					Ok(val) => val,
+					Err(TrapError::Call(subparams, resume)) => {
+						self.kind = CallCreateExecutiveKind::ResumeCall(origin_info, resume, unconfirmed_substate);
+						return Err(TrapError::Call(subparams, self));
+					},
+					Err(TrapError::Create(subparams, address, resume)) => {
+						self.kind = CallCreateExecutiveKind::ResumeCreate(origin_info, resume, unconfirmed_substate);
+						return Err(TrapError::Create(subparams, address, self));
+					},
+				};
+
+				Self::enact_result(&res, state, substate, unconfirmed_substate);
+				Ok(res)
+			},
+			CallCreateExecutiveKind::ResumeCall(..) | CallCreateExecutiveKind::ResumeCreate(..) => panic!("This executive has already been executed once."),
+		}
+	}
+
+	/// Resume execution from a call trap previsouly trapped by `exec`.
+	///
+	/// Current-level tracing is expected to be handled by caller.
+	pub fn resume_call<B: 'a + StateBackend, T: Tracer, V: VMTracer>(mut self, result: vm::MessageCallResult, state: &mut State<B>, substate: &mut Substate, tracer: &mut T, vm_tracer: &mut V) -> ExecutiveTrapResult<'a, FinalizationResult> {
+		match self.kind {
+			CallCreateExecutiveKind::ResumeCall(origin_info, resume, mut unconfirmed_substate) => {
+				let out = {
+					let exec = resume.resume_call(result);
+
+					let mut ext = Self::as_externalities(state, self.info, self.machine, self.schedule, self.depth, self.stack_depth, self.static_flag, &origin_info, &mut unconfirmed_substate, if self.is_create { OutputPolicy::InitContract } else { OutputPolicy::Return }, tracer, vm_tracer);
+					match exec.exec(&mut ext) {
+						Ok(val) => Ok(val.finalize(ext)),
+						Err(err) => Err(err),
+					}
+				};
+
+				let res = match out {
+					Ok(val) => val,
+					Err(TrapError::Call(subparams, resume)) => {
+						self.kind = CallCreateExecutiveKind::ResumeCall(origin_info, resume, unconfirmed_substate);
+						return Err(TrapError::Call(subparams, self));
+					},
+					Err(TrapError::Create(subparams, address, resume)) => {
+						self.kind = CallCreateExecutiveKind::ResumeCreate(origin_info, resume, unconfirmed_substate);
+						return Err(TrapError::Create(subparams, address, self));
+					},
+				};
+
+				Self::enact_result(&res, state, substate, unconfirmed_substate);
+				Ok(res)
+			},
+			CallCreateExecutiveKind::ResumeCreate(..) =>
+				panic!("Resumable as create, but called resume_call"),
```

Comment:
is there any way we can avoid those panics?

Reply:
We can create traits similar to what we do for vm:

```
pub trait Exec: Send {
	fn exec(self: Box<Self>, state: &mut State<B>, ...) -> ExecutiveTrapResult<GasLeft>;
}

pub trait ResumeCall: Send {
	fn resume_call(self: Box<Self>, state: &mut State<B>, ...) -> Box<Exec>;
}

pub trait ResumeCreate: Send {
	fn resume_create(self: Box<Self>, state: &mut State<B>, ...) -> Box<Exec>;
}
```

However, they would need to use traits objects and dynamic dispatch, which is a small amount of runtime overhead. I think it may be a good idea to only use those when in the future we want to expose `Executive` to be some sort of public API.